### ISSUE TITLE
zlib-1.2.11 fails on 68040 and 68060 -O3 but not 68080 -O3

### ISSUE THREAD

**Issue by MBeijer** [bebbo/gcc#115]
gcc version 6.5.0b 190804215041 (GCC) 

**-m68060 -mhard-float -ffast-math -O3:**
```
(insn 1384 772 1385 (set (reg:SI 1 d1 [orig:155 len ] [155])
        (lshiftrt:SI (reg:SI 1 d1 [orig:155 len ] [155])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1384 772 1385 (set (reg:SI 1 d1 [orig:155 len ] [155])
        (lshiftrt:SI (reg:SI 1 d1 [orig:155 len ] [155])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
```

**-m68040 -mhard-float -ffast-math -O3:**
```
(insn 1383 772 1384 (set (reg:SI 1 d1 [orig:155 len ] [155])
        (lshiftrt:SI (reg:SI 1 d1 [orig:155 len ] [155])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@4/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@4/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@4/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1383 772 1384 (set (reg:SI 1 d1 [orig:155 len ] [155])
        (lshiftrt:SI (reg:SI 1 d1 [orig:155 len ] [155])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@4/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@4/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
```

**Comment by MBeijer** [comment]
gcc version 6.5.0b 190818005626 (GCC) 

**-m68040 -mhard-float -O2**
```
(insn 1088 583 1089 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1088 583 1089 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
make[2]: *** [CMakeFiles/zlib.dir/build.make:102: CMakeFiles/zlib.dir/deflate.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
(insn 1088 583 1089 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1088 583 1089 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
```

**-m68060 -mhard-float -O2**
```
(insn 1101 583 1102 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1101 583 1102 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
[ 80%] Building C object CMakeFiles/zlibstatic.dir/uncompr.obj
[ 83%] Building C object CMakeFiles/zlibstatic.dir/zutil.obj
Please submit a full bug report,
with preprocessed source if appropriate.
See <http://gcc.gnu.org/bugs.html> for instructions.
make[2]: *** [CMakeFiles/zlib.dir/build.make:102: CMakeFiles/zlib.dir/deflate.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:110: CMakeFiles/zlib.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
(insn 1101 583 1102 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c: In function 'deflate_stored':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1815:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 1101 583 1102 (set (reg:SI 1 d1 [orig:159 len ] [159])
        (lshiftrt:SI (reg:SI 1 d1 [orig:159 len ] [159])
            (const_int 0 [0]))) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1805 325 {lshrsi3}
     (nil))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master@3/zlib-1.2.11/deflate.c:1815:1: internal compiler error: in final_scan_insn, at final.c:2943
```

**Comment by bebbo** [comment]
`190818005626` is before the fix was commited...

**Comment by MBeijer** [comment]
Oh, didn't notice the fix. Was just supplying more data as I updated to a newer version of the toolchain

**Comment by bebbo** [comment]
-> gcc version 6.5.0b 190819184021

**Comment by bebbo** [comment]
please test

**Comment by MBeijer** [comment]
Will do.

**Comment by MBeijer** [comment]
Works! 
Tested O3 and Os for 68000, 68010, 68020, 68030, 68040, 68060, 68080, 68020-40, 68020-60
all with hard-float and -ffast-math

**Comment by bebbo** [comment]
> 
> 
> **gcc version 6.5.0b 190820204449 (GCC)**
> 
> Works!
> Tested O3 and Os for 68000, 68010, 68020, 68030, 68040, 68060, 68080, 68020-40, 68020-60
> all with hard-float and -ffast-math

it compiles or does it even run some tests with success?

**Comment by MBeijer** [comment]
Let me rephrase, it compiles without issue. Currently building the project it's part of

**Comment by MBeijer** [comment]
Got another insn error on 68040 -O3:
```
(insn 10 727 11 2 (set (reg/v/f:SI 9 a1 [orig:34 control ] [34])
        (mem/f:SI (mem/f/c:SI (reg/v/f:SI 6 d6 [orig:128 argument ] [128]) [50 MEM[(struct png_image_read_control *)argument_11(D)].image+0 S4 A16]) [2 image_13->opaque+0 S4 A16])) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/libpng-1.6.36/pngread.c:3132 39 {*movsi_m68k2}
     (expr_list:REG_EQUIV (mem/f:SI (mem/f/c:SI (reg/v/f:SI 6 d6 [orig:128 argument ] [128]) [50 MEM[(struct png_image_read_control *)argument_11(D)].image+0 S4 A16]) [2 image_13->opaque+0 S4 A16])
        (nil)))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/libpng-1.6.36/pngread.c: In function 'png_image_read_colormapped':
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/libpng-1.6.36/pngread.c:3255:1: error: insn does not satisfy its constraints:
 }
 ^
(insn 10 727 11 2 (set (reg/v/f:SI 9 a1 [orig:34 control ] [34])
        (mem/f:SI (mem/f/c:SI (reg/v/f:SI 6 d6 [orig:128 argument ] [128]) [50 MEM[(struct png_image_read_control *)argument_11(D)].image+0 S4 A16]) [2 image_13->opaque+0 S4 A16])) /opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/libpng-1.6.36/pngread.c:3132 39 {*movsi_m68k2}
     (expr_list:REG_EQUIV (mem/f:SI (mem/f/c:SI (reg/v/f:SI 6 d6 [orig:128 argument ] [128]) [50 MEM[(struct png_image_read_control *)argument_11(D)].image+0 S4 A16]) [2 image_13->opaque+0 S4 A16])
        (nil)))
/opt/bitnami/apps/jenkins/jenkins_home/workspace/DevilutionX_master/libpng-1.6.36/pngread.c:3255:1: internal compiler error: in extract_constrain_insn, at recog.c:2199
```
Doesn't happen on 68060 or 68080

**Comment by MBeijer** [comment]
Btw, the zlib seems to work as intended when tested.

**Comment by MBeijer** [comment]
I've tested zlib, and there's problem reading zlib-zipped data. I've tried 68040, 68060, 68080, O2 and O3 and all of them fail to read. 
Easiest way to test for you is to download https://www.zlib.net/zlib-1.2.11.tar.gz
```
wget https://www.zlib.net/zlib-1.2.11.tar.gz
wget https://raw.githubusercontent.com/AmigaPorts/AmigaCMakeCrossToolchains/master/m68k-amigaos.cmake
tar xvf zlib-1.2.11.tar.gz
cd zlib-1.2.11
mkdir -p build
cd build
cmake .. -DM68K_TOOLCHAIN_PATH=$AMIGAPREFIX  -DCMAKE_INSTALL_PREFIX=$AMIGAPREFIX -DM68K_CPU=68040 -DM68K_FPU=hard -DM68K_COMMON="-O2 -ffast-math" -DCMAKE_TOOLCHAIN_FILE=../../m68k-amigaos.cmake
cmake --build . --config Release --target install -- -j8
```
It should build test example binaries

**Comment by bebbo** [comment]
this works for me to build it:
```
CC=m68k-amigaos-gcc AR=m68k-amigaos-ar RANLIB=m68k-amigaos-ranlib CFLAGS="-O2 -m68040 -noixemul -fomit-frame-pointer -DNO_FSEEKO -ffast-math" ./configure --prefix=/opt/amiga
```
and how do I verify that it's working - or not?

```
vamos -C68020 example
zlib version 1.2.11 = 0x12b0, compile flags = 0x55
uncompress(): hello, hello!
gzread(): hello, hello!
gzgets() after gzseek:  hello!
inflate(): hello, hello!
large_inflate(): OK
after inflateSync(): hello, hello!
inflate with dictionary: hello, hello!
```

**Comment by MBeijer** [comment]
That looks good to me. :)

**Comment by bebbo** [comment]
so why isn't it working with your build?

**Comment by MBeijer** [comment]
It doesn't work as expected in devilutionX, so some further testing than
the demo app might be needed.

**Comment by bebbo** [comment]
ok - I fixed something for #116 and minigzip does now de-/compress correctly. stack size must be sufficient too.

**Comment by bebbo** [comment]
please retest

**Comment by MBeijer** [comment]
This now works properly!
